service/matching/matching_engine.go

```diff
@@ -917,6 +923,73 @@ func (e *matchingEngineImpl) getDefaultBuildId(tqMgr taskQueuePartitionManager)
 	return "", nil
 }
 
+func (e *matchingEngineImpl) getBuildIdTaskReachability(
+	ctx context.Context,
+	nsID,
+	nsName,
+	taskQueue,
+	buildId string,
+) (enumspb.BuildIdTaskReachability, error) {
+	getResp, err := e.GetWorkerVersioningRules(ctx, &matchingservice.GetWorkerVersioningRulesRequest{
+		NamespaceId:                nsID,
+		TaskQueue:                  taskQueue,
+		DeletedRuleInclusionPeriod: durationpb.New(versioningReachabilityDeletedRuleInclusionPeriod),
+	})
+	if err != nil {
+		return enumspb.BUILD_ID_TASK_REACHABILITY_UNSPECIFIED, err
+	}
+	assignmentRules := getResp.GetResponse().GetAssignmentRules()
+	redirectRules := getResp.GetResponse().GetCompatibleRedirectRules()
+
+	// 1. Easy UNREACHABLE case
+	if isTimestampedRedirectRuleSource(buildId, redirectRules) {
+		return enumspb.BUILD_ID_TASK_REACHABILITY_UNREACHABLE, nil
+	}
+
+	// Gather list of all build ids that could point to buildId -> upstreamBuildIds
+	upstreamBuildIds := getUpstreamBuildIds(buildId, redirectRules)
+	buildIdsOfInterest := append(upstreamBuildIds, buildId)
+
+	// 2. Cases for REACHABLE
+	// 2a. If buildId is assignable to new tasks
+	for _, bid := range buildIdsOfInterest {
+		if isReachableAssignmentRuleTarget(bid, assignmentRules) {
+			return enumspb.BUILD_ID_TASK_REACHABILITY_REACHABLE, nil
+		}
+	}
+
+	// 2b. If buildId could be reached from the backlog
+	if existsBacklog, err := existsBackloggedActivityOrWFAssignedToAny(
```

Comment:
let's make it clear that this is about wf tasks, not executions.
```suggestion
	if existsBacklog, err := existsBackloggedActivityOrWFTaskAssignedToAny(
```